backend/telemetry_shard/src/aggregator.rs

```diff
@@ -0,0 +1,271 @@
+    /// can send messages on behalf of more than one chain, and so this ID is
+    /// only really used inside the Aggregator in conjunction with a per-message
+    /// ID.
+    conn_id: AtomicU64,
+    /// Send messages to the aggregator from websockets via this. This is
+    /// stored here so that anybody holding an `Aggregator` handle can
+    /// make use of it.
+    tx_to_aggregator: mpsc::Sender<ToAggregator>,
+}
+
+impl Aggregator {
+    /// Spawn a new Aggregator. This connects to the telemetry backend
+    pub async fn spawn(telemetry_uri: http::Uri) -> anyhow::Result<Aggregator> {
+        let (tx_to_aggregator, rx_from_external) = mpsc::channel(10);
+
+        // Map responses from our connection into messages that will be sent to the aggregator:
+        let tx_from_connection = tx_to_aggregator.clone().with(|msg| {
+            future::ok::<_, mpsc::SendError>(match msg {
+                Message::Connected => ToAggregator::ConnectedToTelemetryCore,
+                Message::Disconnected => ToAggregator::DisconnectedFromTelemetryCore,
+                Message::Data(data) => ToAggregator::FromTelemetryCore(data),
+            })
+        });
+
+        // Establish a resiliant connection to the core (this retries as needed):
+        let tx_to_telemetry_core = create_ws_connection(tx_from_connection, telemetry_uri).await;
+
+        // Handle any incoming messages in our handler loop:
+        tokio::spawn(Aggregator::handle_messages(
+            rx_from_external,
+            tx_to_telemetry_core,
+        ));
+
+        // Return a handle to our aggregator:
+        Ok(Aggregator(Arc::new(AggregatorInternal {
+            conn_id: AtomicU64::new(1),
+            tx_to_aggregator,
+        })))
+    }
+
+    // This is spawned into a separate task and handles any messages coming
+    // in to the aggregator. If nobody is tolding the tx side of the channel
+    // any more, this task will gracefully end.
+    async fn handle_messages(
+        mut rx_from_external: mpsc::Receiver<ToAggregator>,
+        mut tx_to_telemetry_core: mpsc::Sender<FromAggregator>,
+    ) {
+        use internal_messages::{FromShardAggregator, FromTelemetryCore};
+
+        // Just as an optimisation, we can keep track of whether we're connected to the backend
+        // or not, and ignore incoming messages while we aren't.
+        let mut connected_to_telemetry_core = false;
+
+        // A list of close channels for the current connections. Send an empty tuple to
+        // these to ask the connections to be closed.
+        let mut close_connections: Vec<mpsc::Sender<()>> = vec![];
+
+        // Maintain mappings from the connection ID and node message ID to the "local ID" which we
+        // broadcast to the telemetry core.
+        let mut to_local_id = AssignId::new();
+
+        // Any messages coming from nodes that have been muted are ignored:
+        let mut muted: HashSet<ShardNodeId> = HashSet::new();
```

Comment:
I'm not immediately aware of one off the top of my head, although it'd be easy to wrap a collection with some sort of bounding logic.

I had a check and made sure to clean up the muted set when a connection goes away (I wasn't), and made sure to clean up the close_connections list properly (it's now a HashMap).

A node could still cause an issue potentially by broadcasting a large number of message IDs along with a muted chain name, leading to the muted list to keep growing (but probably worse, leading to various other things also growing to account for each node ID).

Reply:
Yeah, a pesky issue. :/

Reply:
One thing I was wondering last night; how many different message IDs do we ever expect to get from a node? I'd assume that we should only ever see a few, so we could always disconnect nodes when we see more than some number to prevent a single node from leading to a lot of state accumulating.

@maciejhirsz do you have any thoughts on this?